lib/x509.mli

```diff
@@ -443,6 +449,196 @@ module Certificate : sig
+      against the [time].  The signature algorithm must be present in
+      [hash_whitelist] (defaults to SHA-2).  The X509v3 extensions of the
+      [chain] are checked, then a chain of trust from [anchors] to the server
+      certificate is validated.  The path length constraints are checked.  The
+      server certificate is checked to contain the given [host], using
+      {!hostnames}.  The returned certificate is the root of the chain, a member
+      of the given list of [anchors]. *)
+  val verify_chain : host:[`host] Domain_name.t option ->
+    time:(unit -> Ptime.t option) ->
+    ?revoked:(issuer:Certificate.t -> cert:Certificate.t -> bool) ->
+    ?hash_whitelist:Mirage_crypto.Hash.hash list ->
+    anchors:(Certificate.t list) -> Certificate.t list ->
+    (Certificate.t, chain_error) result
+
+  (** The polymorphic variant of a fingerprint validation error. *)
+  type fingerprint_validation_error = [
+    | `ServerNameNotPresent of Certificate.t * [`host] Domain_name.t
+    | `NameNotInList of Certificate.t
+    | `InvalidFingerprint of Certificate.t * Cstruct.t * Cstruct.t
+  ]
+
+  (** The polymorphic variant of validation errors. *)
+  type validation_error = [
+    | signature_error
+    | leaf_validation_error
+    | fingerprint_validation_error
+    | `EmptyCertificateChain
+    | `InvalidChain
+  ]
+
+  (** [pp_validation_error ppf validation_error] pretty-prints the
+      [validation_error]. *)
+  val pp_validation_error : validation_error Fmt.t
+
+  type r = ((Certificate.t list * Certificate.t) option, validation_error) result
+
+  (** [verify_chain_of_trust host ~time ~revoked ~hash_whitelist ~anchors certificates]
+      is [result].  First, all possible paths are constructed using the
+      {!build_paths} function, the first certificate of the chain is verified to
+      be a valid leaf certificate (no BasicConstraints extension) and contains
+      the given [host] (using {!hostnames}); if some path is valid, using
+      {!verify_chain}, the result will be [Ok] and contain the actual
+      certificate chain and the trust anchor. *)
+  val verify_chain_of_trust :
+    host:[`host] Domain_name.t option -> time:(unit -> Ptime.t option) ->
+    ?revoked:(issuer:Certificate.t -> cert:Certificate.t -> bool) ->
+    ?hash_whitelist:Mirage_crypto.Hash.hash list ->
+    anchors:(Certificate.t list) -> Certificate.t list -> r
+
+  (** {1 Fingerprint verification} *)
+
+  (** [trust_key_fingerprint host ~time ~hash ~fingerprints certificates] is
+      [result], the first element of [certificates] is verified against the
+      given [fingerprints] list using {!key_fingerprint}.  If [time] is
+      provided, the certificate has to be valid at the given timestamp.  If
+      [host] is provided, the certificate is checked for this name.  The
+      [hostname] of the fingerprint list must match the name in the certificate,
+      using {!hostnames}. *)
+  val trust_key_fingerprint :
```

Comment:
note: this is not new code, only needed to be shuffled around since Validation.signature_error is used in multiple modules.

> extract cert + pubkey fingerprints from certs

isn't [tlsclient](https://github.com/hannesm/tlsclient) the right place for this? it already prints the certificate and public key fingerprint (of the chain).

> does the need for [`host] Domain_name.t prevent using this for *.foo.com wildcard certificates for HTTPS use?

TL;DR: you provide a ``[`host] Domain_name.t * Cstruct.t`` list, and if the server certificate matches on of the Cstruct.t values, it's names (common name / subject alternative name) are checked to include the `` [`host] Domain_name.t `` -- i.e. you specify ("foo.example.com", "00:11:22:33") as authenticator, then if the server certificate fingerprint matches "00:11:22:33", its names are validated against "foo.example.com" -- if it has a subjectAlternativeName with "*.example.com", this is fine :)

Reply:
I think `tlsclient` sounds like a good place. I think people who are looking at this mli string would appreciate a pointer to that. :)